python/tvm/relax/frontend/nn/exporter.py

```diff
@@ -176,35 +183,26 @@ def _unwrap_ret(expr: typing.Any) -> typing.Any:
     def _convert_input(arg):
         if isinstance(arg, tir.Var):
             return rx.Var(arg.name, struct_info=ShapeStructInfo(values=[arg]))
-        if isinstance(arg, (core.Tensor, core.Object)):
+        elif isinstance(arg, (core.Tensor, core.Object)):
             return arg._expr  # pylint: disable=protected-access
-        if isinstance(arg, _spec.Tuple):
+        elif isinstance(arg, _spec.Tuple):
             return rx.Var(
                 arg.name,
                 struct_info=TupleStructInfo(
                     [_convert_input(arg_i).struct_info for arg_i in arg.elements]
                 ),
             )
-        raise TypeError(f"Unsupported input type: {type(arg)}")
+        elif isinstance(arg, rx.Expr):
+            return arg
+        else:
+            raise TypeError(f"Unsupported input type: {type(arg)}")
 
     def _params(mode: str) -> typing.List[rx.Var]:
         inputs: typing.List[rx.Var] = []
 
-        def _get_var(shape_var: tir.Var) -> tir.Var:
-            name = shape_var.name
-            if name in str2var_params:
-                return str2var_params[name]
```

Comment:
removing this would cause issues on handling symbolic shape tying remapping and which allows simplicity of mapping string to variables. We should preserve the behavior

Reply:
Can you describe the behavior that should be preserved?

The main functionality of this step was to ensure symbolic variables are not duplicated across multiple functions, and this functionality is still implemented through `copy_with_new_vars`.  The mapping of strings to variables is still handled [here](https://github.com/apache/tvm/blob/main/python/tvm/relax/frontend/nn/exporter.py#L294), in the `_get_var` function inside `_method_spec_to_input`.

Reply:
Hi @Lunderberg, this PR causes the compilation failure in MLC LLM CI for the Phi model https://ci.mlc.ai/blue/organizations/jenkins/mlc-llm/detail/main/373/

I dug a bit into this. To describe the issue, say the Phi model has hidden size 2560, the final linear layer of the Phi model has weight shape `("vocab_size", 2560)`, and bias shape `("vocab_size",)`.

Prior to this PR, both `"vocab_size"` in the weight and bias are remapped to the same tir.Var, so that for any `x`, `op.matmul(x, op.transpose(weight)) + bias` can have the output shape successfully inferred, represented in `vocab_size`.

Since this PR, the `"vocab_size"` var in the weight and bias shape mismatch. Therefore, the result shape cannot be symbolically inferred, falling back to `R.Tensor(dtype="float16", ndim=3)`, and thus caused the assertion failure as you can see in the CI:

```
  File "/home/ruihang/Workspace/tvm/python/tvm/relax/frontend/nn/modules.py", line 139, in forward
    x = x + self.bias
        ~~^~~~~~~~~~~
  File "/home/ruihang/Workspace/tvm/python/tvm/relax/frontend/nn/_tensor_op.py", line 44, in __add__
    return _op().add(self, other)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ruihang/Workspace/tvm/python/tvm/relax/frontend/nn/op.py", line 105, in add
    return wrap_nested(_op.add(a._expr, b._expr), name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ruihang/Workspace/tvm/python/tvm/relax/frontend/nn/core.py", line 614, in wrap_nested
    return Tensor(_expr=expr)
           ^^^^^^^^^^^^^^^^^^
  File "/home/ruihang/Workspace/tvm/python/tvm/relax/frontend/nn/core.py", line 117, in __init__
    _check_tensor(_expr)
  File "/home/ruihang/Workspace/tvm/python/tvm/relax/frontend/nn/core.py", line 112, in _check_tensor
    assert expr.struct_info.shape is not None
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError  File "/home/ruihang/Workspace/tvm/python/tvm/relax/frontend/nn/modules.py", line 139, in forward
    x = x + self.bias
        ~~^~~~~~~~~~~
  File "/home/ruihang/Workspace/tvm/python/tvm/relax/frontend/nn/_tensor_op.py", line 44, in __add__
    return _op().add(self, other)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ruihang/Workspace/tvm/python/tvm/relax/frontend/nn/op.py", line 105, in add
    return wrap_nested(_op.add(a._expr, b._expr), name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ruihang/Workspace/tvm/python/tvm/relax/frontend/nn/core.py", line 614, in wrap_nested
    return Tensor(_expr=expr)
           ^^^^^^^^^^^^^^^^^^
  File "/home/ruihang/Workspace/tvm/python/tvm/relax/frontend/nn/core.py", line 117, in __init__
    _check_tensor(_expr)
  File "/home/ruihang/Workspace/tvm/python/tvm/relax/frontend/nn/core.py", line 112, in _check_tensor
    assert expr.struct_info.shape is not None
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError
```

Reply:
Thank you for digging into it, and I'll look into it on Monday.

My first impression is that we shouldn't support this behavior while exporting a module.  If we have a dynamic `Linear` module with dynamic shape parameters `input_features` and `output_features`, we should *not* assume that all such instances of the dynamic `Linear` have the same `input_features` and `output_features`.

I think we should normalize all `str` shapes to the same `tir.Var`, but anything already specified as distinct `tir.Var` shapes should remain separate.  This could later be moved to an improved struct inference, which could provide `R.assert_op` on symbolic shapes, rather than falling back to `R.Tensor(ndim=3)`.

Reply:
having ability to have common dynamic variables like `vocab_size` is a pretty common ndeed, so likely we will need to support this behavior.

Reply:
@MasterJH5574 Thank you, and that helps to reproduce the issue.  Does the issue only involve `nn.Linear`, or would it also apply to dynamic shapes that span across multiple independent `nn.Module` subclasses?  If the issue is specific to `nn.Linear`, can you take a look at https://github.com/apache/tvm/pull/16781?  This avoids the duplicate default conversion from `str` to a fresh TIR variable in both the weights and bias.

Overall, I think de-duplicating `str` values makes sense, because Python `str` objects have value equality.  De-duplicating TIR variables by their name doesn't make sense, because TIR variables have reference equality.  I think we could have a good intermediate implementation if we de-duplicate `str` values prior to conversion into TIR variables.

Reply:
@MasterJH5574 Can you test against https://github.com/apache/tvm/pull/16785?  It re-implements the same functionality that I require on my side, while maintaining the handling of dynamic shapes specified by strings.